src/main/scala/loamstream/model/jobs/TerminationReason.scala

```diff
@@ -1,50 +1,64 @@
 package loamstream.model.jobs
 
+import scala.reflect.ClassTag
+import loamstream.util.Classes
+
 /**
  * @author clint
  * Apr 16, 2019
  * 
  * An ADT for the reasons why a job was killed by the underlying system (Uger/LSF/Google) 
  */
-sealed trait TerminationReason
+sealed abstract class TerminationReason(
+    val raw: Option[String], 
+    companion: TerminationReason.Companion[_]) {
+  
+  import TerminationReason._
+  
+  final def name: String = companion.Name
+  
+  final def isRunTime: Boolean = this.isInstanceOf[RunTime]
+  final def isCpuTime: Boolean = this.isInstanceOf[CpuTime]
+  final def isMemory: Boolean = this.isInstanceOf[Memory]
+  final def isUserRequested: Boolean = this.isInstanceOf[UserRequested]
+  final def isUnclassified: Boolean = this.isInstanceOf[Unclassified]
+  final def isUnknown: Boolean = this.isInstanceOf[Unknown]
+}
 
 object TerminationReason {
+  abstract class Companion[A : ClassTag] {
+    val Name: String = Classes.simpleNameOf[A]
+  }
+  
   //A was killed because it ran too long 
-  case object RunTime extends TerminationReason
+  final case class RunTime(override val raw: Option[String] = None) extends TerminationReason(raw, RunTime)
```

Comment:
I'm also curious what having the companion object as a member of the class accomplishes? Is it just for the name of it? It's probably more complicated than it's worth, but I have to believe Scala has a way of getting the companion object at runtime via reflection. 

If so, a little more complexity here to avoid a copy/paste error here as there's no validating that the companion object passed to the constructor actually is the companion object of this class.

Probably 6 vs. 1/2 dozen. I'd lean more towards the reflection and avoiding copy/paste errors. But I know you prefer to be more explicit and have run into reflection issues in the past.

Reply:
> I'm also curious what having the companion object as a member of the class accomplishes? Is it just for the name of it? It's probably more complicated than it's worth, but I have to believe Scala has a way of getting the companion object at runtime via reflection.

I wanted to encode the names once (as companion object names, in this case) instead of once in a class def and then again in `fromNameAndRawValue`.  A downside is as you mentioned, there's no compile-time guarantee that the companion object passed to a class's constructor "matches" the class. 

> Probably 6 vs. 1/2 dozen. I'd lean more towards the reflection and avoiding copy/paste errors. But I know you prefer to be more explicit and have run into reflection issues in the past.

I'll see what I can come up with.  `TerminationReason` may change (see earlier comments), which make some of the current setup obsolete.  I'd very much like to prevent copy-paste errors, _and_ avoid the zillion edge cases that come from reflection.